src/Storages/MaterializedView/RefreshTimers.cpp

```diff
@@ -0,0 +1,243 @@
+{
+    if (time_interval)
+    {
+        for (auto && [kind, value] : time_interval->kinds)
+            setWithKind(kind, value);
+    }
+}
+
+std::chrono::sys_seconds RefreshAfterTimer::after(std::chrono::system_clock::time_point tp) const
+{
+    auto tp_date = std::chrono::floor<std::chrono::days>(tp);
+    auto tp_time_offset = std::chrono::floor<std::chrono::seconds>(tp - tp_date);
+    std::chrono::year_month_day ymd(tp_date);
+    ymd += years;
+    ymd += months;
+    std::chrono::sys_days date = ymd;
+    date += weeks;
+    date += days;
+    auto result = std::chrono::time_point_cast<std::chrono::seconds>(date);
+    result += tp_time_offset;
+    result += hours;
+    result += minutes;
+    result += seconds;
+    return result;
+}
+
+void RefreshAfterTimer::setWithKind(IntervalKind kind, UInt64 val)
+{
+    switch (kind)
+    {
+        case IntervalKind::Second:
+            seconds = std::chrono::seconds{val};
+            break;
+        case IntervalKind::Minute:
+            minutes = std::chrono::minutes{val};
+            break;
+        case IntervalKind::Hour:
+            hours = std::chrono::hours{val};
+            break;
+        case IntervalKind::Day:
+            days = std::chrono::days{val};
+            break;
+        case IntervalKind::Week:
+            weeks = std::chrono::weeks{val};
+            break;
+        case IntervalKind::Month:
+            months = std::chrono::months{val};
+            break;
+        case IntervalKind::Year:
+            years = std::chrono::years{val};
+            break;
+        default:
+            break;
+    }
+}
+
+RefreshEveryTimer::RefreshEveryTimer(const ASTTimePeriod & time_period, const ASTTimeInterval * time_offset)
+    : offset(time_offset)
+    , value{static_cast<UInt32>(time_period.value)}
+    , kind{time_period.kind}
+{
+    // TODO: validate invariants
+}
+
+std::chrono::sys_seconds RefreshEveryTimer::next(std::chrono::system_clock::time_point tp) const
+{
+    if (value == 0)
+        return std::chrono::floor<std::chrono::seconds>(tp);
+    switch (kind)
+    {
+        case IntervalKind::Second:
+            return alignedToSeconds(tp);
+        case IntervalKind::Minute:
+            return alignedToMinutes(tp);
+        case IntervalKind::Hour:
+            return alignedToHours(tp);
+        case IntervalKind::Day:
+            return alignedToDays(tp);
+        case IntervalKind::Week:
+            return alignedToWeeks(tp);
+        case IntervalKind::Month:
+            return alignedToMonths(tp);
+        case IntervalKind::Year:
+            return alignedToYears(tp);
+        default:
+            return std::chrono::ceil<std::chrono::seconds>(tp);
+    }
+}
+
+std::chrono::sys_seconds RefreshEveryTimer::alignedToYears(std::chrono::system_clock::time_point tp) const
+{
+    using namespace std::chrono_literals;
+
+    auto tp_days = std::chrono::floor<std::chrono::days>(tp);
```

Comment:
It is really tricky one. System time doesn't have leap seconds and utc_now - sys_now is not a zero. Furthermore utc time is not equal to tai, so it isn't clear which one to prefer. At the same time system clock is simple to read by human, because it represents days/months/years as a round number and has a built in time zone, so I preferred it. Probably it can become a trouble in case of geo distributed scheduling, but it is "out of scope" problem

Reply:
Ah, system_clock pauses for leap seconds, TIL! Should've checked, my bad.